cellrank/tools/_utils.py

```diff
@@ -1403,3 +1414,247 @@ def _colors_in_order(
         color_list.append(adata.uns[f"{cluster_key}_colors"][mask][0])
 
     return color_list
+
+
+def _get_cat_and_null_indices(
+    cat_series: Series,
+) -> Tuple[np.ndarray, np.ndarray, Dict[Any, np.ndarray]]:
+    """
+    Given a categorical :class:`pandas.Series`, get the indices corresponding to categories and NaNs.
+
+    Params
+    ------
+    cat_series
+        Series that contains categorical annotations.
+
+    Returns
+    -------
+    :class: `numpy.ndarray`
+        Array containing the indices of elements corresponding to categories in `cat_series`.
+    :class: `numpy.ndarray`
+        Array containing the indices of elements corresponding to NaNs in `cat_series`.
+    :class:`dict`
+        Dict containing categories of `cat_series` as keys and an array of corresponding indices as values.
+    """
+
+    # check the dtype
+    if cat_series.dtype != "category":
+        raise TypeError(
+            f"Expected `cat_series` to be categorical, found `{cat_series.dtype.name!r}`."
+        )
+
+    # define a dict that has category names as keys and arrays of indices as values
+    lookup_dict = {
+        cat: np.where(cat_series == cat)[0] for cat in cat_series.cat.categories
+    }
+    all_indices = np.arange(len(cat_series))
+
+    # collect all category indices
+    cat_indices = np.concatenate(list(lookup_dict.values()))
+
+    # collect all null indices (the ones where we have NaN in `cat_series`)
+    null_indices = np.array(list(set(all_indices) - set(cat_indices)))
+
+    # check that null indices and cat indices are unique
+    assert (
+        np.unique(cat_indices, return_counts=True)[1] == 1
+    ).all(), "Cat indices are not unique."
+    assert (
+        np.unique(null_indices, return_counts=True)[1] == 1
+    ).all(), "Null indices are not unique."
+
+    # check that there is no overlap
+    assert (
+        len(set(cat_indices).intersection(set(null_indices))) == 0
+    ), "Cat and null indices overlap."
+
+    # check that their untion is the set of all indices
+    assert set(cat_indices).union(set(null_indices)) == set(
+        all_indices
+    ), "Some indices got lost on the way."
+
+    return cat_indices, null_indices, lookup_dict
+
+
+def _solve_lin_system(
+    mat_a: Union[np.ndarray, spmatrix],
+    mat_b: Union[np.ndarray, spmatrix],
+    solver: str = "direct",
+    tol: float = 1e-5,
+    related_columns_in_b: Optional[Iterable] = None,
+    use_eye: bool = False,
```

Comment:
@Marius1311 I've added this here because there was a code repetition in the caller in BaseEstimator. I hope you approve.

Reply:
yes, I do.